collections/src/main/scala/strawman/collection/immutable/ChampHashMap.scala

```diff
@@ -0,0 +1,680 @@
+  *
+  *  @tparam K      the type of the keys contained in this hash set.
+  *  @tparam V      the type of the values associated with the keys in this hash map.
+  *
+  *  @author  Michael J. Steindorfer
+  *  @version 2.13
+  *  @since   2.13
+  *  @define Coll `immutable.ChampHashMap`
+  *  @define coll immutable champ hash map
+  */
+
+@SerialVersionUID(2L)
+private[immutable] sealed case class ChampHashMap[K, +V](val rootNode: MapNode[K, V], val cachedJavaHashCode: Int, val cachedSize: Int)
+  extends Map[K, V]
+    with MapOps[K, V, ChampHashMap, ChampHashMap[K, V]]
+    with StrictOptimizedIterableOps[(K, V), Iterable /* ChampHashMap */, ChampHashMap[K, V]]
+    with Serializable {
+
+  override def iterableFactory = List
+
+  override def mapFactory = ChampHashMap
+
+  override protected[this] def mapFromIterable[K2, V2](it: collection.Iterable[(K2, V2)]): ChampHashMap[K2, V2] = ChampHashMap.from(it)
+
+  override protected[this] def fromSpecificIterable(coll: collection.Iterable[(K, V)]): ChampHashMap[K, V] = ChampHashMap.from(coll)
+
+  override protected[this] def newSpecificBuilder(): Builder[(K, V), ChampHashMap[K, V]] = ChampHashMap.newBuilder()
+
+  override def knownSize: Int = cachedSize
+
+  override def iterator(): Iterator[(K, V)] = new MapKeyValuePairIterator[K, V](rootNode)
+
+  override def get(key: K): Option[V] = rootNode.get(key, computeHash(key), 0)
+
+  override def updated[V1 >: V](key: K, value: V1): ChampHashMap[K, V1] = {
+    val effect = MapEffect[K, V1]()
+    val keyHash = computeHash(key)
+    val newRootNode = rootNode.updated(key, value, keyHash, 0, effect)
+
+    if (effect.isModified) {
+      if (effect.hasReplacedValue) {
+        ChampHashMap(newRootNode, cachedJavaHashCode, cachedSize)
+      } else {
+        ChampHashMap(newRootNode, cachedJavaHashCode + keyHash, cachedSize + 1)
+      }
+    }
+    else this
+  }
+
+  override def remove(key: K): ChampHashMap[K, V] = {
+    val effect = MapEffect[K, V]()
+    val keyHash = computeHash(key)
+    val newRootNode = rootNode.removed(key, keyHash, 0, effect)
+
+    if (effect.isModified)
+      ChampHashMap(newRootNode, cachedJavaHashCode - keyHash, cachedSize - 1)
+    else this
+  }
+
+  override def empty: ChampHashMap[K, V] = ChampHashMap.empty
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case map: ChampHashMap[K, V] =>
+        (this eq map) ||
+          (this.cachedSize == map.cachedSize) &&
+            (this.cachedJavaHashCode == map.cachedJavaHashCode) &&
+            (this.rootNode == map.rootNode)
+      case _ => super.equals(that)
+    }
+
+  override def hashCode(): Int = collection.Set.unorderedHash(toIterable, "Map".##)
+
+}
+
+object MapNode {
+
+  private val EmptyMapNode = new BitmapIndexedMapNode(0, 0, Array.empty)
+
+  def empty[K, V]: MapNode[K, V] =
+    EmptyMapNode.asInstanceOf[MapNode[K, V]]
+
+  val TupleLength = 2
```

Comment:
We should not expose these constants publicly. They could also be made into compile time constants.

```
private final val TupleLength = 2 // final, no type annotation, callers will constant fold this in.
```

Reply:
Thanks. I "finalized" all the constants. (I assumed that `val`s in companion objects are final anyways, but seems that I wrong there.)

I agree in not exposing these constants, but then it would suffice to make the companion object private (e.g., `private[immutable] object MapNode {}`) instead of each constant, right?